lib/dsc-lib/src/functions/cidr_host.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use super::Function;
+use crate::configure::context::Context;
+use crate::functions::{FunctionArgKind, FunctionCategory, FunctionMetadata};
+use crate::DscError;
+use ipnetwork::{IpNetwork, Ipv4Network, Ipv6Network};
+use rust_i18n::t;
+use serde_json::Value;
+use tracing::debug;
+
+#[derive(Debug, Default)]
+pub struct CidrHost {}
+
+impl Function for CidrHost {
+    fn get_metadata(&self) -> FunctionMetadata {
+        FunctionMetadata {
+            name: "cidrHost".to_string(),
+            description: t!("functions.cidrHost.description").to_string(),
+            category: vec![FunctionCategory::String],
+            min_args: 2,
+            max_args: 2,
+            accepted_arg_ordered_types: vec![
+                vec![FunctionArgKind::String],
+                vec![FunctionArgKind::Number],
+            ],
+            remaining_arg_accepted_types: None,
+            return_types: vec![FunctionArgKind::String],
+        }
+    }
+
+    fn invoke(&self, args: &[Value], _context: &Context) -> Result<Value, DscError> {
+        debug!("{}", t!("functions.cidrHost.invoked"));
+
+        let cidr_string = args[0].as_str().ok_or_else(|| {
+            DscError::FunctionArg(
+                "cidrHost".to_string(),
+                t!("functions.cidrHost.invalidNetwork").to_string(),
+            )
+        })?;
+
+        let host_index = args[1].as_i64().ok_or_else(|| {
+            DscError::FunctionArg(
+                "cidrHost".to_string(),
+                t!("functions.cidrHost.invalidHostIndex").to_string(),
+            )
+        })?;
+
+        if host_index < 0 {
+            return Err(DscError::FunctionArg(
+                "cidrHost".to_string(),
+                t!("functions.cidrHost.negativeHostIndex").to_string(),
+            ));
+        }
+
+        let network = cidr_string.parse::<IpNetwork>().map_err(|_| {
+            DscError::FunctionArg(
+                "cidrHost".to_string(),
+                t!("functions.cidrHost.invalidCidr", cidr = cidr_string).to_string(),
+            )
+        })?;
+
+        let result = match network {
+            IpNetwork::V4(net) => calculate_ipv4_host(net, host_index as u32)?,
+            IpNetwork::V6(net) => calculate_ipv6_host(net, host_index as u128)?,
+        };
+
+        Ok(Value::String(result))
+    }
+}
+
+fn calculate_ipv4_host(net: Ipv4Network, host_index: u32) -> Result<String, DscError> {
```

Comment:
```suggestion
fn calculate_ipv4_host(net: &Ipv4Network, host_index: u32) -> Result<String, DscError> {
```

Should use a reference for structs so a copy isn't implicitly made